lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,3148 @@
+  /// which they are used (see `contractAllocsPass`). This works outside-in.
+  /// Lowering virtual qubit allocations opens up more qubit reuse opportunities
+  /// within inner scopes, and gives the remaining optimizations more
+  /// flexibility.
+  ///
+  /// Then, an inside-out analysis/optimization pass is performed, assigning
+  /// physical qubits to virtual wires, and lifting common optimizations. This
+  /// inside-out analysis/optimization pass works as follows:
+  /// - Step 1: `if`s inside blocks are resolved (hence inside-out), once an
+  /// `if` is resolved, the information contained by it will not be changed
+  /// (with the possible exception of re-indexing physical qubits).
+  /// - Step 2: blocks are resolved.
+  /// * First the nodes inside them are scheduled, assigning a cycle to every
+  ///   node such that each node is scheduled after all of its dependencies, and
+  ///   before all of its successors.
+  /// * Then, physical qubits are allocated and assigned to virtual wires based
+  ///   on lifetime information (i.e., which cycles the virtual wire is used
+  ///   in). This algorithm treats `if`s as "solid rectangles", where all qubits
+  ///   in use anywhere in the `if` are considered in use for the entire `if` by
+  ///   the parent scope. In other words, the lifetime analysis does not look
+  ///   inside `if`s.
+  /// - Step 3: Once its blocks are resolved, then the parent `if` is resolved.
+  /// * First, the allocations from the two inner blocks are combined/matched
+  ///   (respecting reuse within the blocks but with re-indexing allowed).
+  /// * Second, equivalent operations at the beginning/end of the then and else
+  ///   blocks are lifted to the parent context, before/after the `if`.
+  /// - Step 4: return to Step 1 for the parent block the `if` is in, or Step 2
+  ///   if all `if`s in the block have been resolved.
+  void runOnOperation() override {
+    auto mod = getOperation();
+
+    for (auto &op : mod) {
+      if (auto func = dyn_cast<func::FuncOp>(op)) {
+        if (!func->hasAttr("cudaq-kernel"))
+          continue;
+
+        if (func.getNumResults() == 0) {
+          func->emitRemark("Function marked 'cudaq-kernel' returns no results, "
+                           "qubit management skipped.");
+          continue;
+        }
+
+        validateFunc(func);
+
+        Block *oldBlock = &func.front();
+
+        auto engine = DependencyAnalysisEngine();
+
+        // Construct a DependencyBlock for the function body based on the quake
+        // AST
+        auto body = engine.visitBlock(
+            oldBlock, SmallVector<DependencyNode::DependencyEdge>());
+
+        if (!body) {
+          LLVM_DEBUG(
+              llvm::dbgs()
+              << "Encountered issue, backing out and skipping function\n");
+          engine.clean();
+          continue;
+        }
+
+        OpBuilder builder(func);
+        LifeTimeAnalysis set(cudaq::opt::topologyAgnosticWiresetName);
```

Comment:
First the final assignment of wires is determined (by performAnalysis), and only in the last step is new quake code generated. Also, because of `assign-wire-indices`, it should be `borrow_wire`s in and `borrow_wire`s out.